### PR TITLE
perf(python): bound response content-type inspection

### LINKED ISSUE #33313: perf(python): bound response Content-Type inspection before decoding
## Problem

`has_event_stream_media_type()` decodes and copies an entire upstream Content-Type value to inspect only the media type before the first semicolon. This work runs synchronously in the mitmproxy response-header hook and is not covered by the existing WebSocket header-inspection budget.

Audited at `5e019e4f0a8ee96c89df9f636e0749c8beda9828`; the affected function and stream-setup caller are also unchanged at current-main snapshot `4b4f5d498a2c024852cd7694f5f13a722d177432`.

## Location and cause

- [`crates/runner/mitm-addon/src/http_response_classification.py:28-32`](https://github.com/vm0-ai/vm0/blob/5e019e4f0a8ee96c89df9f636e0749c8beda9828/crates/runner/mitm-addon/src/http_response_classification.py#L28-L32): `Headers.get()` converts matching raw values to strings before any size check; `partition(";")` then materializes the otherwise unused parameter suffix.
- [`crates/runner/mitm-addon/src/response_streaming.py:196-224`](https://github.com/vm0-ai/vm0/blob/5e019e4f0a8ee96c89df9f636e0749c8beda9828/crates/runner/mitm-addon/src/response_streaming.py#L196-L224): billable model-provider response setup invokes this classification synchronously. The terminal JSON fallback also invokes it at lines 102-110.
- [`crates/runner/mitm-addon/src/mitm_addon.py:1601-1612`](https://github.com/vm0-ai/vm0/blob/5e019e4f0a8ee96c89df9f636e0749c8beda9828/crates/runner/mitm-addon/src/mitm_addon.py#L1601-L1612): the response-header hook calls stream setup directly.

The pinned mitmproxy 12.2.3 HTTP/1 response-head parser accepts the oversized field. Its `Headers.get_all()` uses UTF-8 decoding with `surrogateescape`, so a non-UTF-8 parameter suffix expands to a two-byte-per-character Python string, followed by another suffix copy. No response-body data or network-body capture is needed to trigger this cost.

## Reproduction and measured impact

In the locked CPython 3.14.0 / mitmproxy 12.2.3 environment, parse a Content-Type field containing `text/event-stream; pad="` followed by repeated `0xff` bytes and a closing quote, then invoke the real `mitm_addon.responseheaders()` hook for an eligible OpenAI Responses flow.

The local reproduction produced:

| Parameter bytes | Hook peak temporary allocation | Hook time with tracemalloc |
| --- | ---: | ---: |
| 0 | 7,264 bytes | 0.295 ms |
| 1 MiB | 4,198,152 bytes | 8.931 ms |
| 8 MiB | 33,558,264 bytes | 73.164 ms |

Input construction and HTTP-head parsing were completed before measurement. In every case the SSE parser was installed and the original raw header remained unchanged. A spy on mitmproxy's byte-to-string conversion boundary confirmed conversion of the full 8,388,633-byte value. Independently, the classifier alone took a median 61.787 ms without tracemalloc over five calls for the 8 MiB non-UTF-8 case. An ASCII suffix also produces linear extra copies: approximately 16 MiB of peak allocation for an 8 MiB value.

This is an adversarial oversized-upstream-header case, not an observed production incident or a claim that ordinary provider headers are this large. It adds avoidable allocation and event-loop latency on top of mitmproxy's existing raw HTTP/1 head buffering. Fixing this classifier alone will not bound every other header consumer or the dependency's head buffer.

Minimal reproduction, run from `crates/runner/mitm-addon` after setting up the locked environment:

```bash
uv run --no-sync python - <<'PY'
import gc
import tracemalloc
from mitmproxy import http
from mitmproxy.net.http import http1
from mitmproxy.test import tflow
import flow_metadata_keys as keys
import mitm_addon

value = b'text/event-stream; pad="' + b'\xff' * (8 * 1024 * 1024) + b'"'
flow = tflow.tflow()
flow.request = http.Request.make("POST", "https://api.openai.com/v1/responses", b"{}")
flow.response = http1.read_response_head([
    b"HTTP/1.1 200 OK", b"Content-Type: " + value, b"Content-Length: 2",
])
flow.metadata.update({
    keys.FIREWALL_NAME: "model-provider:openai-api-key",
    keys.CLI_AGENT_TYPE: "codex",
    keys.FIREWALL_BILLABLE: True,
    keys.MODEL_USAGE_PROVIDER: "gpt-5.5",
    keys.ORIGINAL_URL: "https://api.openai.com/v1/responses",
})
gc.collect()
tracemalloc.start()
mitm_addon.responseheaders(flow)
print("peak_bytes:", tracemalloc.get_traced_memory()[1])
tracemalloc.stop()
assert "model_sse_usage_finish" in flow.metadata
assert flow.response.headers.fields[0][1] == value
PY
```

## Acceptance criteria

- Inspect raw header fields with an explicit work/allocation bound before converting values to strings; do not decode or copy an irrelevant parameter suffix just to classify the leading media type.
- Preserve ordinary case-insensitive SSE detection, optional whitespace and parameters, and the existing rejection of `text/event-stream+json` and `application/json; profile="text/event-stream"` lookalikes. Specify safe behavior when the classification budget is exhausted.
- Preserve original request/response wire headers. Keep usage and failure observers on the same classification contract.
- Add structural regression tests through the response-header hook for oversized ASCII and non-UTF-8 values, guarding the conversion boundary instead of relying on timing thresholds. Include missing/repeated fields and budget-exhaustion cases.
- Retain existing parser-setup and shared failure/usage classification tests. The fresh audit's 234 sampled tests and 602 related tests currently pass, so the reproduction exposes a coverage gap rather than an existing test failure.

Open-issue title searches for Content-Type, media type, response/header, and Python performance found no duplicate before filing.

### PR DESCRIPTION
**Pull request by seven332**
An upstream `Content-Type: text/event-stream; ...` with a large parameter suffix previously caused the synchronous response-header hook to decode and copy the entire value. Classify the raw singleton field within limits of 8,192 fields and 8,192 prefix bytes, and normalize only the complete media type before the first in-budget semicolon.

Ordinary SSE detection, optional SP/HTAB, case-insensitive matching, and large irrelevant parameter suffixes remain supported. Missing/repeated fields and incomplete over-budget prefixes consistently return non-SSE through the existing shared usage/failure classifier. This intentionally removes the semicolon-dependent behavior of duplicate Content-Type fields. Headers and streamed bytes remain unchanged; unclassifiable true SSE responses may lack parsed usage/failure events. The bound is local to this classifier and does not cover mitmproxy's raw head buffering or other header consumers. No wire protocol or persisted-state changes are involved.

Closes #33313.

## Validation

- New raw-header regression suite: 12 failed / 8 passed against the original classifier; all 20 passed after the fix. Covers real HTTP/1 ASCII and non-UTF-8 parameter suffixes, dependency conversion guards, parsed usage, duplicate fields, and exact/over-budget boundaries without timing thresholds.
- **308 related tests passed** in the locked CPython 3.14.0 / mitmproxy 12.2.3 environment, including shared failure reporting, parser setup, SSE accounting, stream lifecycle, response encoding, WebSocket lifecycle, and error cleanup:

  ```bash
  uv run --no-sync python -m pytest tests/test_response_content_type_budget.py tests/test_model_provider_response_parser_setup.py tests/test_model_provider_failure_reporting.py tests/test_response_headers_handler.py tests/test_response_handler_cleanup.py tests/test_response_stream_buffering.py tests/test_response_stream_state_release.py tests/test_response_encoding_inspection_risk.py tests/test_model_provider_sse_usage_openai_responses.py tests/test_model_provider_sse_usage_anthropic.py tests/test_model_provider_websocket_lifecycle.py tests/test_error_handler.py -q --tb=short
  ```

- `uv lock --check`, `uv run --no-sync ruff format --check .`, `uv run --no-sync ruff check .`, and `uv run --no-sync basedpyright -p .` passed.
- `npx --yes --package=prettier@3.8.1 prettier --check docs/mitm-addon-contracts.md`, staged diff whitespace checks, and file-size checks passed.
- Full repository checks are left to the PR pipeline; no local Vitest or development server was run.

### CODE REVIEW

File: crates/runner/mitm-addon/tests/test_response_content_type_budget.py [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+    def lower(self) -> bytes:
+        raise AssertionError("classifier normalized the complete parameterized value")
+
+    def find(self, sub, start=0, end=None) -> int:
+        assert end is not None
+        assert 0 <= start <= end <= _PREFIX_LIMIT
+        return super().find(sub, start, end)
+
+    @overload
+    def __getitem__(self, key: SupportsIndex) -> int: ...
```
[the remark this answers is not in the document]
Comment (seven332) [reply crates/runner/mitm-addon/tests/test_response_content_type_budget.py:181]: This ellipsis is the body of a typing.overload declaration, which specifies the integer-index return type for type checkers. The concrete __getitem__ implementation immediately below performs the bounded access check and returns the byte. This follows the existing overload pattern in test_http_header_syntax.py; BasedPyright and all relevant tests pass. No runtime statement is missing, so this diagnostic is a false positive.

File: crates/runner/mitm-addon/tests/test_response_content_type_budget.py [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+    @overload
+    def __getitem__(self, key: slice) -> bytes: ...
```
[the remark this answers is not in the document]
Comment (seven332) [reply crates/runner/mitm-addon/tests/test_response_content_type_budget.py:184]: This is the slice overload paired with the integer-index overload above. The ellipsis is an intentional typing.overload stub; the following concrete __getitem__ implementation handles slices and enforces the prefix-copy bound. The overload preserves the bytes return type for slicing and is validated by BasedPyright. This diagnostic does not identify a missing runtime operation.

**Review by seven332 — COMMENTED** [review]
LGTM

**Reviewed head:** `fb33b28411de46ce3a626cb798717c9b2b5f3fb2`
**Main practice revision:** `b938cc661c0ba74101f744580411a5d73f39996f`

**Summary:** Bounds raw response Content-Type field enumeration and media-type prefix inspection without decoding or copying irrelevant parameter suffixes. Missing, duplicate, and exhausted inputs consistently use the existing shared non-SSE classification; ordinary SSE, original headers, and streamed bytes are preserved.

**Findings:** No actionable findings after the complete submitted diff, callers, and six-pass self-review. The two static-analysis diagnostics on `typing.overload` ellipsis stubs are false positives; both have evidence-backed replies and are resolved. No new compatibility fallback or persisted/wire contract is introduced.

**Testing:** 308 relevant locked-environment addon tests passed, including the real HTTP/1 oversized ASCII/non-UTF-8 regressions, structural allocation guards, exact boundaries, parsed usage, shared failure reporting, and response lifecycle coverage. Ruff format/check, BasedPyright, lock validation, Markdown formatting, local document links, diff whitespace, and file-size checks passed. No local Vitest or development server was run.

The limit is specific to this classifier; other header consumers and mitmproxy's raw head buffering are outside it. Unclassifiable true SSE responses may lack parsed usage/failure events, as documented. CI and repository protection remain separate gates. This is a self-authored review, posted as COMMENT rather than GitHub approval.

**Comment by seven332** [comment]
## Code Review: PR #33392

**Reviewed head:** `fb33b28411de46ce3a626cb798717c9b2b5f3fb2`
**Verdict:** LGTM

The [commit-bound review](https://github.com/vm0-ai/vm0/pull/33392#pullrequestreview-5174458022) records the complete review against main practices at `b938cc661c0ba74101f744580411a5d73f39996f`.

**Summary:** Bounded raw singleton/prefix inspection preserves ordinary SSE and large irrelevant parameter suffixes while consistently handling duplicate/exhausted input. Both usage and failure observers retain the same classification contract.

**Findings:** No actionable findings after all six self-review passes. Both static-analysis threads on intentional `typing.overload` stubs were explained and resolved without code changes.

**Testing:** 308 relevant addon tests and locked Ruff/BasedPyright checks passed; lockfile, Markdown, local links, diff whitespace, and file-size checks passed. Full repository CI is a separate gate. Raw head buffering and other header consumers remain outside this classifier's guarantee; unclassifiable true SSE responses can lack parsed usage/failure events.

This marker and the COMMENT review do not supply an independent GitHub approval or authorize merge.
